assignment3/GridWordSearcher.java

```diff
@@ -0,0 +1,121 @@
+package assignment3;
+
+import java.util.*;
+
+public class GridWordSearcher {
+    private int n;
+    private int m;
+    private char[][] grid;
+    private WordDictionary dict;
+    private static List<List<Integer>> directions;
+
+    static {
+        directions = new ArrayList<>();
+        for (int i = -1; i <= 1; i++) {
+            for (int j = -1; j <= 1; j++) {
+                if (i == 0 && j == 0) {
+                    continue;
+                }
+                directions.add(Arrays.asList(i, j));
+            }
+        }
+    }
+
+    private static class GridPosition {
+        private int i;
+        private int j;
+        private int direction;
+
+        GridPosition(int i, int j, int direction) {
+            this.i = i;
+            this.j = j;
+            this.direction = direction;
+        }
+
+        int getI() {
+            return i;
+        }
+
+        int getJ() {
+            return j;
+        }
+
+        int getDirection() {
+            return direction;
+        }
+
+        void setDirection(int direction) {
+            this.direction = direction;
+        }
+    }
+
+    public GridWordSearcher(int n, int m, char[][] grid, WordDictionary dict) {
+        this.n = n;
+        this.m = m;
+        this.grid = grid;
+        this.dict = dict;
+    }
+
+    public Set<String> findWords() {
+        boolean[][] marks = new boolean[n][m];
+        Set<String> result = new HashSet<>();
+        for (int i = 0; i < n; i++) {
+            for (int j = 0; j < m; j++) {
+                findWordsFromPosition(i, j, marks, result);
+            }
+        }
+        return result;
+    }
+
+    private void findWordsFromPosition(int i, int j, boolean[][]marks, Set<String> result) {
+        String firstLetterString = Character.toString(grid[i][j]);
+        if (!dict.isPrefix(firstLetterString)) {
+            return;
+        }
+        if (dict.isWord(firstLetterString)) {
+            result.add(firstLetterString);
+        }
+        StringBuilder currentPrefix = new StringBuilder(firstLetterString);
+        List<GridPosition> stack = new ArrayList<>();
+        stack.add(new GridPosition(i, j, 0));
+        marks[i][j] = true;
+
+        // dfs from the current cell
+        while (!stack.isEmpty()) {
+            GridPosition top = stack.get(stack.size() - 1);
+            boolean isStillTop = true;
+
+            // if there are any unvisited neighbours
+            if (top.getDirection() < directions.size()) {
+                int nextI = top.getI() + directions.get(top.getDirection()).get(0);

```

Comment:
This is an ideal opportunity to show you can write object oriented code (in an interview). You could have two classes here:
class Position {
  int row;
  int column;

 /** @throw IllegalArgumentException if position not valid
  public void validate(int maxRow, int maxColumn) {
    ...
  }
}
class Step {
  int rowStep;
  int columnStep;

  /** Applies this step to a position to create a new position */
  public Position step(Position original) {
    ...
  }
}

Then this code would look like
 int nextPosition = top.getDirection().step(top.getPosition());

Try when writing code to not just write like you were writing C, and take advantage of the fact you have an object oriented language to make your code more readable if you are writing java.